File: ios/loveaceios/Service/YKTService.swift
import Foundation
import os

private let logger = Logger(subsystem: "tech.loveace.loveaceios", category: "YKTService")

actor YKTService {
    private let connection: AUFEConnection
    static let baseURL = "http://ykt-aufe-edu-cn-s.vpn2.aufe.edu.cn:8118"

    init(connection: AUFEConnection) {
        self.connection = connection
    }

    func initSession() async -> UniResponse<Void> {
        do {
            let client = await connection.client!
            _ = try await client.get("\(Self.baseURL)/casLogin.jsp")
            return .success(())
        } catch {
            logger.error("initSession failed: \(error.localizedDescription)")
            return .failure(error.localizedDescription)
        }
    }

    func getBalance() async -> UniResponse<CardBalance> {
        do {
            let client = await connection.client!
            let (data, _) = try await client.get("\(Self.baseURL)/queryUserBalances.action")
            let html = String(data: data, encoding: .utf8) ?? ""
            let patterns = [
                "余额[：:]?\\s*</label>\\s*<label>\\s*([\\d.]+)\\s*元",
                "余额[：:]?\\s*([\\d.]+)\\s*元"
            ]
            var balanceStr: String?
            for pattern in patterns {
                if let regex = try? NSRegularExpression(pattern: pattern, options: .caseInsensitive),
                   let match = regex.firstMatch(in: html, range: NSRange(html.startIndex..., in: html)),
                   let range = Range(match.range(at: 1), in: html) {
                    balanceStr = String(html[range])
                    break
                }
            }
            guard let bs = balanceStr, let balance = Double(bs) else {
                throw ServiceError.parseError("无法解析余额")
            }
            return .success(CardBalance(balance: balance, balanceText: "\(bs)元"))
        } catch {
            logger.error("getBalance failed: \(error.localizedDescription)")
            return .failure(error.localizedDescription, retryable: true)
        }
    }

    func getTransactions(startDate: String, endDate: String) async -> UniResponse<[TransactionRecord]> {
        do {
            let client = await connection.client!
            let (data, _) = try await client.post("\(Self.baseURL)/queryUserCostList.action",
                                                   formData: ["startDate": startDate, "endDate": endDate])
            let html = String(data: data, encoding: .utf8) ?? ""
            let records = parseTransactionHTML(html)
            return .success(records)
        } catch {
            logger.error("getTransactions failed: \(error.localizedDescription)")
            return .failure(error.localizedDescription, retryable: true)
        }
    }

    private func parseTransactionHTML(_ html: String) -> [TransactionRecord] {
        var records: [TransactionRecord] = []
        let pattern = "<tr>\\s*<td[^>]*>(.*?)</td>\\s*<td[^>]*>(.*?)</td>\\s*<td[^>]*>(.*?)</td>\\s*<td[^>]*>(.*?)</td>\\s*<td[^>]*>(.*?)</td>\\s*<td[^>]*>(.*?)</td>\\s*<td[^>]*>(.*?)</td>\\s*<td[^>]*>(.*?)</td>\\s*</tr>"
        guard let regex = try? NSRegularExpression(pattern: pattern, options: [.caseInsensitive, .dotMatchesLineSeparators]) else { return [] }
        let matches = regex.matches(in: html, range: NSRange(html.startIndex..., in: html))
        for match in matches {
            var groups: [String] = []
            for i in 1...8 {
                guard let range = Range(match.range(at: i), in: html) else { groups.append(""); continue }
                var val = String(html[range])
                val = val.replacingOccurrences(of: "<[^>]*>", with: "", options: .regularExpression)
                    .replacingOccurrences(of: "&nbsp;", with: " ").trimmingCharacters(in: .whitespaces)
                groups.append(val)
            }
            let numericOnly = { (s: String) -> Double? in
                Double(s.replacingOccurrences(of: "[^\\d.\\-]", with: "", options: .regularExpression))
            }
            records.append(TransactionRecord(
                accountingTime: groups[0], transactionTime: groups[1],
                expense: numericOnly(groups[2]), income: numericOnly(groups[3]),
                operationType: groups[4], balance: numericOnly(groups[5]) ?? 0.0,
                area: groups[6], terminalId: groups[7]
            ))
        }
        return records
    }

    // MARK: - Electricity Payment

    func getPageInfo() async -> UniResponse<StudentInfo> {
        do {
            let client = await connection.client!
            let (data, _) = try await client.get("\(Self.baseURL)/utilityUnBindUserPowerPayInit.action")
            let html = String(data: data, encoding: .utf8) ?? ""
            return .success(StudentInfo.fromHTML(html))
        } catch {
            logger.error("getPageInfo failed: \(error.localizedDescription)")
            return .failure(error.localizedDescription)
        }
    }

    func getDormList() async -> UniResponse<[SelectOption]> { await getOptions(dormId: "", buildingId: "", floorId: "", dormName: "") }
    func getBuildingList(dormId: String, dormName: String) async -> UniResponse<[SelectOption]> { await getOptions(dormId: dormId, buildingId: "", floorId: "", dormName: dormName) }
    func getFloorList(dormId: String, buildingId: String, dormName: String) async -> UniResponse<[SelectOption]> { await getOptions(dormId: dormId, buildingId: buildingId, floorId: "", dormName: dormName) }
    func getRoomList(dormId: String, buildingId: String, floorId: String, dormName: String) async -> UniResponse<[SelectOption]> { await getOptions(dormId: dormId, buildingId: buildingId, floorId: floorId, dormName: dormName) }

    private func getOptions(dormId: String, buildingId: String, floorId: String, dormName: String) async -> UniResponse<[SelectOption]> {
        do {
            let client = await connection.client!
            let (data, _) = try await client.post("\(Self.baseURL)/utilitBindXiaoQuData.action",
                                                   formData: ["dormId": dormId, "buildingId": buildingId, "floorId": floorId, "dormName": dormName])
            let body = String(data: data, encoding: .utf8) ?? ""
            return .success(SelectOption.parseList(body))
        } catch {
            logger.error("getOptions failed: \(error.localizedDescription)")
            return .failure(error.localizedDescription)
        }
    }

    func payElectricity(request: UtilityPaymentRequest) async -> UniResponse<UtilityPaymentResult> {
        do {
            let client = await connection.client!
            logger.info("payElectricity request: \(request.toFormData())")
            let (data, _) = try await client.post("\(Self.baseURL)/utilityUnBindUserPowerPay.action", formData: request.toFormData())
            let html = String(data: data, encoding: .utf8) ?? ""
            logger.info("payElectricity response length: \(html.count)")
            logger.info("payElectricity response preview: \(String(html.prefix(500)))")
            let result = UtilityPaymentResult.fromHTML(html)
            logger.info("payElectricity result: success=\(result.success), message=\(result.message)")
            return .success(result)
        } catch {
            logger.error("payElectricity failed: \(error.localizedDescription)")
            return .failure(error.localizedDescription)
        }
    }

    func getPurchaseHistory(startDate: String, endDate: String) async -> UniResponse<ElectricPurchaseQueryResult> {
        do {
            let client = await connection.client!
            let (data, _) = try await client.post("\(Self.baseURL)/utilityQueryRunningAccountInfo.action",
                                                   formData: ["startDate": "\(startDate) 00:00:00", "endDate": "\(endDate) 23:59:59"])
            let html = String(data: data, encoding: .utf8) ?? ""
            return .success(Self.parsePurchaseHistory(html: html, startDate: startDate, endDate: endDate))
        } catch {
            logger.error("getPurchaseHistory failed: \(error.localizedDescription)")
            return .failure(error.localizedDescription)
        }
    }

    private static func parsePurchaseHistory(html: String, startDate: String, endDate: String) -> ElectricPurchaseQueryResult {
        var records: [ElectricPurchaseRecord] = []
        let pattern = "<tr>\\s*<td>([^<]*)</td>\\s*<td>([^<]*)</td>\\s*<td>([^<]*)</td>\\s*<td[^>]*>([^<]*)</td>\\s*<td>([^<]*)</td>\\s*<td[^>]*>([^<]*)</td>\\s*<td>([^<]*)</td>\\s*</tr>"
        guard let regex = try? NSRegularExpression(pattern: pattern, options: [.caseInsensitive, .dotMatchesLineSeparators]) else {
            return ElectricPurchaseQueryResult(startDate: startDate, endDate: endDate, records: [])
        }
        let matches = regex.matches(in: html, range: NSRange(html.startIndex..., in: html))
        for match in matches {
            var groups: [String] = []
            for i in 1...7 {
                guard let range = Range(match.range(at: i), in: html) else { groups.append(""); continue }
                groups.append(String(html[range]).trimmingCharacters(in: .whitespaces))
            }
            guard groups[0] != "姓名", !groups[0].isEmpty else { continue }
            records.append(ElectricPurchaseRecord(
                name: groups[0], studentId: groups[1], area: groups[2], roomInfo: groups[3],
                amount: Double(groups[4]) ?? 0.0, purchaseDate: groups[5], department: groups[6]
            ))
        }
        return ElectricPurchaseQueryResult(startDate: startDate, endDate: endDate, records: records)
    }
}

File: android/app/src/main/java/tech/loveace/appv3/data/service/YKTService.kt
package tech.loveace.appv3.data.service

import android.util.Log
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import tech.loveace.appv3.data.model.*
import tech.loveace.appv3.data.network.AUFEConnection

/**
 * 一卡通系统服务 - 余额查询、消费记录
 */
class YKTService(private val connection: AUFEConnection) {

    suspend fun initSession(): UniResponse<Unit> = withContext(Dispatchers.IO) {
        try {
            connection.client.get("$BASE_URL/casLogin.jsp")
            UniResponse.success(Unit)
        } catch (e: Exception) {
            Log.e(TAG, "initSession failed", e)
            UniResponse.failure(e.message ?: "初始化一卡通会话失败")
        }
    }

    suspend fun getBalance(): UniResponse<CardBalance> = withContext(Dispatchers.IO) {
        try {
            val response = connection.client.get("$BASE_URL/queryUserBalances.action")
            val html = response.body?.string() ?: throw Exception("响应为空")
            val balanceMatch = Regex("余额[：:]?\\s*</label>\\s*<label>\\s*([\\d.]+)\\s*元", RegexOption.IGNORE_CASE).find(html)
                ?: Regex("余额[：:]?\\s*([\\d.]+)\\s*元", RegexOption.IGNORE_CASE).find(html)
                ?: throw Exception("无法解析余额")
            val balance = balanceMatch.groupValues[1].toDoubleOrNull() ?: 0.0
            UniResponse.success(CardBalance(balance, "${balanceMatch.groupValues[1]}元"))
        } catch (e: Exception) {
            Log.e(TAG, "getBalance failed", e)
            UniResponse.failure(e.message ?: "查询余额失败", retryable = true)
        }
    }

    suspend fun getTransactions(startDate: String, endDate: String): UniResponse<List<TransactionRecord>> =
        withContext(Dispatchers.IO) {
            try {
                Log.d(TAG, "💳 正在查询消费记录: $startDate ~ $endDate")
                val response = connection.client.post(
                    "$BASE_URL/queryUserCostList.action",
                    formData = mapOf("startDate" to startDate, "endDate" to endDate),
                )
                val html = response.body?.string() ?: throw Exception("响应为空")
                val records = parseTransactionHtml(html)
                Log.d(TAG, "💳 消费记录查询成功: 共${records.size}条")
                UniResponse.success(records)
            } catch (e: Exception) {
                Log.e(TAG, "getTransactions failed", e)
                UniResponse.failure(e.message ?: "查询消费记录失败", retryable = true)
            }
        }

    private fun parseTransactionHtml(html: String): List<TransactionRecord> {
        val records = mutableListOf<TransactionRecord>()
        val rowRegex = Regex(
            "<tr>\\s*<td[^>]*>(.*?)</td>\\s*<td[^>]*>(.*?)</td>\\s*<td[^>]*>(.*?)</td>\\s*<td[^>]*>(.*?)</td>\\s*<td[^>]*>(.*?)</td>\\s*<td[^>]*>(.*?)</td>\\s*<td[^>]*>(.*?)</td>\\s*<td[^>]*>(.*?)</td>\\s*</tr>",
            setOf(RegexOption.IGNORE_CASE, RegexOption.DOT_MATCHES_ALL)
        )
        for (match in rowRegex.findAll(html)) {
            try {
                val g = match.groupValues.drop(1).map { it.replace(Regex("<[^>]*>"), "").replace("&nbsp;", " ").trim() }
                records.add(TransactionRecord(
                    accountingTime = g[0], transactionTime = g[1],
                    expense = g[2].replace(Regex("[^\\d.\\-]"), "").toDoubleOrNull(),
                    income = g[3].replace(Regex("[^\\d.\\-]"), "").toDoubleOrNull(),
                    operationType = g[4],
                    balance = g[5].replace(Regex("[^\\d.\\-]"), "").toDoubleOrNull() ?: 0.0,
                    area = g[6], terminalId = g[7],
                ))
            } catch (_: Exception) { continue }
        }
        return records
    }

    companion object {
        private const val TAG = "YKTService"
        const val BASE_URL = "http://ykt-aufe-edu-cn-s.vpn2.aufe.edu.cn:8118"
    }

    // ── 电费充值相关 ──

    suspend fun getPageInfo(): UniResponse<StudentInfo> = withContext(Dispatchers.IO) {
        try {
            val response = connection.client.get("$BASE_URL/utilityUnBindUserPowerPayInit.action")
            val html = response.body?.string() ?: throw Exception("响应为空")
            UniResponse.success(StudentInfo.fromHtml(html))
        } catch (e: Exception) {
            Log.e(TAG, "getPageInfo failed", e)
            UniResponse.failure(e.message ?: "获取学生信息失败")
        }
    }

    suspend fun getDormList(): UniResponse<List<SelectOption>> = getOptions("", "", "", "")
    suspend fun getBuildingList(dormId: String, dormName: String): UniResponse<List<SelectOption>> = getOptions(dormId, "", "", dormName)
    suspend fun getFloorList(dormId: String, buildingId: String, dormName: String): UniResponse<List<SelectOption>> = getOptions(dormId, buildingId, "", dormName)
    suspend fun getRoomList(dormId: String, buildingId: String, floorId: String, dormName: String): UniResponse<List<SelectOption>> = getOptions(dormId, buildingId, floorId, dormName)

    private suspend fun getOptions(dormId: String, buildingId: String, floorId: String, dormName: String): UniResponse<List<SelectOption>> =
        withContext(Dispatchers.IO) {
            try {
                val response = connection.client.post(
                    "$BASE_URL/utilitBindXiaoQuData.action",
                    formData = mapOf("dormId" to dormId, "buildingId" to buildingId, "floorId" to floorId, "dormName" to dormName),
                )
                val body = response.body?.string() ?: ""
                UniResponse.success(SelectOption.parseList(body))
            } catch (e: Exception) {
                Log.e(TAG, "getOptions failed", e)
                UniResponse.failure(e.message ?: "获取选项失败")
            }
        }

    suspend fun payElectricity(request: UtilityPaymentRequest): UniResponse<UtilityPaymentResult> =
        withContext(Dispatchers.IO) {
            try {
                val response = connection.client.post(
                    "$BASE_URL/utilityUnBindUserPowerPay.action",
                    formData = request.toFormData(),
                )
                val html = response.body?.string() ?: throw Exception("响应为空")
                UniResponse.success(UtilityPaymentResult.fromHtml(html))
            } catch (e: Exception) {
                Log.e(TAG, "payElectricity failed", e)
                UniResponse.failure(e.message ?: "充值失败")
            }
        }

    suspend fun getPurchaseHistory(startDate: String, endDate: String): UniResponse<ElectricPurchaseQueryResult> =
        withContext(Dispatchers.IO) {
            try {
                val response = connection.client.post(
                    "$BASE_URL/utilityQueryRunningAccountInfo.action",
                    formData = mapOf("startDate" to "$startDate 00:00:00", "endDate" to "$endDate 23:59:59"),
                )
                val html = response.body?.string() ?: throw Exception("响应为空")
                UniResponse.success(ElectricPurchaseQueryResult.fromHtml(html, startDate, endDate))
            } catch (e: Exception) {
                Log.e(TAG, "getPurchaseHistory failed", e)
                UniResponse.failure(e.message ?: "获取购电记录失败")
            }
        }
}
